src/differentials.jl

```diff
@@ -206,3 +206,27 @@ end
 end
 
 Base.conj(x::Thunk) = @thunk(conj(extern(x)))
+
+
+####
+"""
+    NO_FIELDS
+
+Constant for the reverse-mode derivative with respect to a structure that has no fields.
+The most notable use for this is for the reverse-mode derivative with respect to the
+function itself, when that function is not a closure.
+"""
+const NO_FIELDS = DNE()
+
+####
+"""
+    differential(𝒟::Type, der)
```

Comment:
Maybe this should take `primal` and `conjugate` as arguments, and depending on `𝒟` return either `Wirtinger` or their sum? I think that would make it more clear to rule authors, that when you create a `Wirtinger`, you usually also want this fall-through behavior.

Reply:
It could also be useful to check, whether `conjugate isa Zero` here, and unwrap `Wirtinger` if that's the case.

Reply:
>Maybe this should take primal and conjugate as arguments, and depending on 𝒟 return either Wirtinger or their sum?

I'm not sure I understand the advantag of that?

With this we have
`differential(𝒟, Wirtinger(primal, conjugatge))`
which seems fine.
What is the advantage of
`differential(𝒟, primal, conjugatge)` 
?

> It could also be useful to check, whether conjugate isa Zero here, and unwrap Wirtinger if that's the case.

Maybe.
Maybe even `iszero(conjugate)` to to get constants like `0`

Maybe even for inputs that are scalar:
`iszero(der) && return Zero()`
and similar for `One()`

We should discuss that kind thing in an issue and make a follow up PR for it

Reply:
> What is the advantage of
> `differential(𝒟, primal, conjugatge)`
> ?

I just don't know if `differential` is the best name for this function, since it takes a differential and returns a, sometimes different and more efficient, differential again. I would expect a function called `differential` to work more like a constructor. Or do we maybe want to call this `wirtinger` and have it take `𝒟`, `primal`, and `conjugate`, since this probably corresponds better to what it does right now? But I also wouldn't feel too strongly just leaving this for now, since I'm also struggling to find a better name for this function.

> Maybe.
> Maybe even `iszero(conjugate)` to to get constants like `0`
> 
> Maybe even for inputs that are scalar:
> `iszero(der) && return Zero()`
> and similar for `One()`

I'm not quite convinced we benefit from introducing dispatch based on value here, wouldn't this also cause problem on GPUs? But this is definitely an issue for another day.

Reply:
How about we rename it to `refine_differential` ?
I think this actually also interacts with #8  since we will want to apply something recursively.